{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Experiment\n",
    "This notebook generates the experiment results from output using both GEMMA and linear regression in the directory output. The complete list of SNP-s and the p-value computed is generated for each algorithm and a graphical representation of the SNP-s is outputted as PNG.\n",
    "\n",
    "Generating all the output for all phenotypes would take ~3 days (ofc it can be easily parallelized by running different phenotypes in parallel). The GEMMA algorithm is optimized hence runs much faster than the linear regression.\n",
    "\n",
    "An example output is in example_output.tar.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Ensure R kernel is installed\n",
    "# !conda install -c r r-irkernel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "The downloaded source packages are in\n",
      "\t‘/tmp/RtmpoPBq7o/downloaded_packages’\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Updating HTML index of packages in '.Library'\n",
      "Making 'packages.html' ... done\n",
      "Warning message:\n",
      ": dependency ‘calibrate’ is not availableWarning message:\n",
      "In install.packages(\"qqman\", repos = \"http://cran.us.r-project.org\"): installation of package ‘qqman’ had non-zero exit status"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "The downloaded source packages are in\n",
      "\t‘/tmp/RtmpoPBq7o/downloaded_packages’\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Updating HTML index of packages in '.Library'\n",
      "Making 'packages.html' ... done\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "The downloaded source packages are in\n",
      "\t‘/tmp/RtmpoPBq7o/downloaded_packages’\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Updating HTML index of packages in '.Library'\n",
      "Making 'packages.html' ... done\n"
     ]
    }
   ],
   "source": [
    "install.packages(\"qtl\", repos = \"http://cran.us.r-project.org\")\n",
    "install.packages(\"qqman\", repos = \"http://cran.us.r-project.org\")\n",
    "install.packages(\"data.table\", repos = \"http://cran.us.r-project.org\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Map QTLs for phenotypes measured in CFW outbred mice using the linear\n",
    "# mixed model (LMM) analysis implemented in GEMMA.\n",
    "library(qtl)\n",
    "library(data.table)\n",
    "source(\"src/misc.R\")\n",
    "source(\"src/gemma.R\")\n",
    "source(\"src/read.data.R\")\n",
    "source(\"src/data.manip.R\")\n",
    "source(\"src/qtl.analyses.R\")\n",
    "\n",
    "# SCRIPT PARAMETERS\n",
    "# -----------------\n",
    "chromosomes    <- NULL\n",
    "gemmadir       <- \"./gemma\"\n",
    "gemma.exe      <- paste(getwd(),\"/gemma-0.98.4-linux-static-AMD64\",sep=\"\")\n",
    "\n",
    "geno_txt       <- \"data/geno.txt\"\n",
    "map_txt        <- \"data/map.txt\"\n",
    "pheno_csv      <- \"data/pheno.csv\"\n",
    "\n",
    "# LOAD PHENOTYPE DATA\n",
    "# -------------------\n",
    "# Load the phenotype data, and discard outlying phenotype values. I\n",
    "# create binary covariates from some of the categorical phenotypes.\n",
    "cat(\"Loading phenotype data.\\n\")\n",
    "pheno_all <- read.pheno(pheno_csv)\n",
    "pheno_all <- prepare.pheno(pheno_all)\n",
    "pheno_all <- cbind(pheno_all,\n",
    "               binary.from.categorical(pheno_all$FCbox,paste0(\"FCbox\",1:4)),\n",
    "               binary.from.categorical(pheno_all$PPIbox,paste0(\"PPIbox\",1:5)),\n",
    "               binary.from.categorical(pheno_all$methcage,\n",
    "                                       paste0(\"methcage\",1:12)),\n",
    "               binary.from.categorical(pheno_all$round,paste0(\"SW\",1:25)))\n",
    "\n",
    "\n",
    "# LOAD GENOTYPE DATA\n",
    "# ------------------\n",
    "# Load the \"mean genotypes\"; i.e., the the mean alternative allele\n",
    "# counts.\n",
    "cat(\"Loading genotype data.\\n\")\n",
    "map     <- read.map(map_txt)\n",
    "out     <- read.geno.dosage(geno_txt,nrow(map))\n",
    "discard <- out$discard\n",
    "X_all   <- out$geno\n",
    "rm(out)\n",
    "\n",
    "# Discard genotype samples from mislabeled flowcell samples.\n",
    "X_all <- X_all[which(discard == \"no\"),]\n",
    "\n",
    "\n",
    "# Discard SNPs with low \"imputation quality\" assessed by inspecting\n",
    "# the genotype probabilities. Retain SNPs for which: (1) at least 95%\n",
    "# of the samples have a maximum probability genotype greater than than\n",
    "# 0.5; (2) the minor allele frequency is greater than 2%.\n",
    "f       <- apply(X_all,2,compute.maf)\n",
    "markers <- which(map$quality > 0.95 & f > 0.02)\n",
    "map     <- map[markers,]\n",
    "X_all   <- X_all[,markers]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A matrix: 30 × 10 of type dbl</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>cfw-1-3207478</th><th scope=col>cfw-1-4592184</th><th scope=col>rs31954814</th><th scope=col>rs31947195</th><th scope=col>rs30660852</th><th scope=col>rs260800880</th><th scope=col>rs49005848</th><th scope=col>rs31616579</th><th scope=col>rs31668389</th><th scope=col>rs31445079</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>26305</th><td>0</td><td>0</td><td>2.000</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td></tr>\n",
       "\t<tr><th scope=row>26306</th><td>0</td><td>0</td><td>0.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td></tr>\n",
       "\t<tr><th scope=row>26307</th><td>0</td><td>2</td><td>0.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td></tr>\n",
       "\t<tr><th scope=row>26308</th><td>0</td><td>0</td><td>0.776</td><td>2.000</td><td>2.000</td><td>2.000</td><td>0.968</td><td>2.000</td><td>0.973</td><td>2.000</td></tr>\n",
       "\t<tr><th scope=row>26309</th><td>0</td><td>0</td><td>0.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td></tr>\n",
       "\t<tr><th scope=row>26310</th><td>0</td><td>1</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td></tr>\n",
       "\t<tr><th scope=row>26312</th><td>0</td><td>0</td><td>2.000</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td></tr>\n",
       "\t<tr><th scope=row>26313</th><td>0</td><td>0</td><td>0.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td></tr>\n",
       "\t<tr><th scope=row>26314</th><td>0</td><td>2</td><td>1.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td></tr>\n",
       "\t<tr><th scope=row>26315</th><td>0</td><td>0</td><td>1.439</td><td>1.000</td><td>1.561</td><td>2.000</td><td>1.000</td><td>1.559</td><td>1.000</td><td>0.561</td></tr>\n",
       "\t<tr><th scope=row>26316</th><td>0</td><td>0</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td></tr>\n",
       "\t<tr><th scope=row>26317</th><td>0</td><td>0</td><td>0.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td></tr>\n",
       "\t<tr><th scope=row>26318</th><td>0</td><td>0</td><td>1.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td></tr>\n",
       "\t<tr><th scope=row>26319</th><td>0</td><td>0</td><td>0.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td></tr>\n",
       "\t<tr><th scope=row>26320</th><td>0</td><td>0</td><td>0.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td></tr>\n",
       "\t<tr><th scope=row>26321</th><td>0</td><td>0</td><td>1.031</td><td>2.000</td><td>2.000</td><td>2.000</td><td>1.031</td><td>2.000</td><td>1.031</td><td>2.000</td></tr>\n",
       "\t<tr><th scope=row>26322</th><td>0</td><td>0</td><td>0.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td></tr>\n",
       "\t<tr><th scope=row>26323</th><td>0</td><td>0</td><td>0.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td></tr>\n",
       "\t<tr><th scope=row>26324</th><td>0</td><td>0</td><td>2.000</td><td>1.077</td><td>1.077</td><td>1.077</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td></tr>\n",
       "\t<tr><th scope=row>26325</th><td>0</td><td>0</td><td>2.000</td><td>1.000</td><td>1.000</td><td>1.000</td><td>1.000</td><td>1.000</td><td>1.000</td><td>1.000</td></tr>\n",
       "\t<tr><th scope=row>26326</th><td>0</td><td>0</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td></tr>\n",
       "\t<tr><th scope=row>26327</th><td>0</td><td>0</td><td>0.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td></tr>\n",
       "\t<tr><th scope=row>26328</th><td>0</td><td>0</td><td>0.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td></tr>\n",
       "\t<tr><th scope=row>26329</th><td>0</td><td>0</td><td>2.000</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td></tr>\n",
       "\t<tr><th scope=row>26330</th><td>0</td><td>0</td><td>2.000</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td></tr>\n",
       "\t<tr><th scope=row>26331</th><td>0</td><td>0</td><td>0.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td></tr>\n",
       "\t<tr><th scope=row>26332</th><td>0</td><td>0</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td></tr>\n",
       "\t<tr><th scope=row>26333</th><td>0</td><td>0</td><td>1.039</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td></tr>\n",
       "\t<tr><th scope=row>26334</th><td>0</td><td>2</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td></tr>\n",
       "\t<tr><th scope=row>26335</th><td>0</td><td>1</td><td>0.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td><td>2.000</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 30 × 10 of type dbl\n",
       "\\begin{tabular}{r|llllllllll}\n",
       "  & cfw-1-3207478 & cfw-1-4592184 & rs31954814 & rs31947195 & rs30660852 & rs260800880 & rs49005848 & rs31616579 & rs31668389 & rs31445079\\\\\n",
       "\\hline\n",
       "\t26305 & 0 & 0 & 2.000 & 0.000 & 0.000 & 0.000 & 0.000 & 0.000 & 0.000 & 0.000\\\\\n",
       "\t26306 & 0 & 0 & 0.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000\\\\\n",
       "\t26307 & 0 & 2 & 0.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000\\\\\n",
       "\t26308 & 0 & 0 & 0.776 & 2.000 & 2.000 & 2.000 & 0.968 & 2.000 & 0.973 & 2.000\\\\\n",
       "\t26309 & 0 & 0 & 0.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000\\\\\n",
       "\t26310 & 0 & 1 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000\\\\\n",
       "\t26312 & 0 & 0 & 2.000 & 0.000 & 0.000 & 0.000 & 0.000 & 0.000 & 0.000 & 0.000\\\\\n",
       "\t26313 & 0 & 0 & 0.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000\\\\\n",
       "\t26314 & 0 & 2 & 1.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000\\\\\n",
       "\t26315 & 0 & 0 & 1.439 & 1.000 & 1.561 & 2.000 & 1.000 & 1.559 & 1.000 & 0.561\\\\\n",
       "\t26316 & 0 & 0 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000\\\\\n",
       "\t26317 & 0 & 0 & 0.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000\\\\\n",
       "\t26318 & 0 & 0 & 1.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000\\\\\n",
       "\t26319 & 0 & 0 & 0.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000\\\\\n",
       "\t26320 & 0 & 0 & 0.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000\\\\\n",
       "\t26321 & 0 & 0 & 1.031 & 2.000 & 2.000 & 2.000 & 1.031 & 2.000 & 1.031 & 2.000\\\\\n",
       "\t26322 & 0 & 0 & 0.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000\\\\\n",
       "\t26323 & 0 & 0 & 0.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000\\\\\n",
       "\t26324 & 0 & 0 & 2.000 & 1.077 & 1.077 & 1.077 & 2.000 & 2.000 & 2.000 & 2.000\\\\\n",
       "\t26325 & 0 & 0 & 2.000 & 1.000 & 1.000 & 1.000 & 1.000 & 1.000 & 1.000 & 1.000\\\\\n",
       "\t26326 & 0 & 0 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000\\\\\n",
       "\t26327 & 0 & 0 & 0.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000\\\\\n",
       "\t26328 & 0 & 0 & 0.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000\\\\\n",
       "\t26329 & 0 & 0 & 2.000 & 0.000 & 0.000 & 0.000 & 0.000 & 0.000 & 0.000 & 0.000\\\\\n",
       "\t26330 & 0 & 0 & 2.000 & 0.000 & 0.000 & 0.000 & 0.000 & 0.000 & 0.000 & 0.000\\\\\n",
       "\t26331 & 0 & 0 & 0.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000\\\\\n",
       "\t26332 & 0 & 0 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000\\\\\n",
       "\t26333 & 0 & 0 & 1.039 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000\\\\\n",
       "\t26334 & 0 & 2 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000\\\\\n",
       "\t26335 & 0 & 1 & 0.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000 & 2.000\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 30 × 10 of type dbl\n",
       "\n",
       "| <!--/--> | cfw-1-3207478 | cfw-1-4592184 | rs31954814 | rs31947195 | rs30660852 | rs260800880 | rs49005848 | rs31616579 | rs31668389 | rs31445079 |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 26305 | 0 | 0 | 2.000 | 0.000 | 0.000 | 0.000 | 0.000 | 0.000 | 0.000 | 0.000 |\n",
       "| 26306 | 0 | 0 | 0.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 |\n",
       "| 26307 | 0 | 2 | 0.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 |\n",
       "| 26308 | 0 | 0 | 0.776 | 2.000 | 2.000 | 2.000 | 0.968 | 2.000 | 0.973 | 2.000 |\n",
       "| 26309 | 0 | 0 | 0.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 |\n",
       "| 26310 | 0 | 1 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 |\n",
       "| 26312 | 0 | 0 | 2.000 | 0.000 | 0.000 | 0.000 | 0.000 | 0.000 | 0.000 | 0.000 |\n",
       "| 26313 | 0 | 0 | 0.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 |\n",
       "| 26314 | 0 | 2 | 1.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 |\n",
       "| 26315 | 0 | 0 | 1.439 | 1.000 | 1.561 | 2.000 | 1.000 | 1.559 | 1.000 | 0.561 |\n",
       "| 26316 | 0 | 0 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 |\n",
       "| 26317 | 0 | 0 | 0.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 |\n",
       "| 26318 | 0 | 0 | 1.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 |\n",
       "| 26319 | 0 | 0 | 0.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 |\n",
       "| 26320 | 0 | 0 | 0.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 |\n",
       "| 26321 | 0 | 0 | 1.031 | 2.000 | 2.000 | 2.000 | 1.031 | 2.000 | 1.031 | 2.000 |\n",
       "| 26322 | 0 | 0 | 0.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 |\n",
       "| 26323 | 0 | 0 | 0.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 |\n",
       "| 26324 | 0 | 0 | 2.000 | 1.077 | 1.077 | 1.077 | 2.000 | 2.000 | 2.000 | 2.000 |\n",
       "| 26325 | 0 | 0 | 2.000 | 1.000 | 1.000 | 1.000 | 1.000 | 1.000 | 1.000 | 1.000 |\n",
       "| 26326 | 0 | 0 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 |\n",
       "| 26327 | 0 | 0 | 0.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 |\n",
       "| 26328 | 0 | 0 | 0.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 |\n",
       "| 26329 | 0 | 0 | 2.000 | 0.000 | 0.000 | 0.000 | 0.000 | 0.000 | 0.000 | 0.000 |\n",
       "| 26330 | 0 | 0 | 2.000 | 0.000 | 0.000 | 0.000 | 0.000 | 0.000 | 0.000 | 0.000 |\n",
       "| 26331 | 0 | 0 | 0.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 |\n",
       "| 26332 | 0 | 0 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 |\n",
       "| 26333 | 0 | 0 | 1.039 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 |\n",
       "| 26334 | 0 | 2 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 |\n",
       "| 26335 | 0 | 1 | 0.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 | 2.000 |\n",
       "\n"
      ],
      "text/plain": [
       "      cfw-1-3207478 cfw-1-4592184 rs31954814 rs31947195 rs30660852 rs260800880\n",
       "26305 0             0             2.000      0.000      0.000      0.000      \n",
       "26306 0             0             0.000      2.000      2.000      2.000      \n",
       "26307 0             2             0.000      2.000      2.000      2.000      \n",
       "26308 0             0             0.776      2.000      2.000      2.000      \n",
       "26309 0             0             0.000      2.000      2.000      2.000      \n",
       "26310 0             1             2.000      2.000      2.000      2.000      \n",
       "26312 0             0             2.000      0.000      0.000      0.000      \n",
       "26313 0             0             0.000      2.000      2.000      2.000      \n",
       "26314 0             2             1.000      2.000      2.000      2.000      \n",
       "26315 0             0             1.439      1.000      1.561      2.000      \n",
       "26316 0             0             2.000      2.000      2.000      2.000      \n",
       "26317 0             0             0.000      2.000      2.000      2.000      \n",
       "26318 0             0             1.000      2.000      2.000      2.000      \n",
       "26319 0             0             0.000      2.000      2.000      2.000      \n",
       "26320 0             0             0.000      2.000      2.000      2.000      \n",
       "26321 0             0             1.031      2.000      2.000      2.000      \n",
       "26322 0             0             0.000      2.000      2.000      2.000      \n",
       "26323 0             0             0.000      2.000      2.000      2.000      \n",
       "26324 0             0             2.000      1.077      1.077      1.077      \n",
       "26325 0             0             2.000      1.000      1.000      1.000      \n",
       "26326 0             0             2.000      2.000      2.000      2.000      \n",
       "26327 0             0             0.000      2.000      2.000      2.000      \n",
       "26328 0             0             0.000      2.000      2.000      2.000      \n",
       "26329 0             0             2.000      0.000      0.000      0.000      \n",
       "26330 0             0             2.000      0.000      0.000      0.000      \n",
       "26331 0             0             0.000      2.000      2.000      2.000      \n",
       "26332 0             0             2.000      2.000      2.000      2.000      \n",
       "26333 0             0             1.039      2.000      2.000      2.000      \n",
       "26334 0             2             2.000      2.000      2.000      2.000      \n",
       "26335 0             1             0.000      2.000      2.000      2.000      \n",
       "      rs49005848 rs31616579 rs31668389 rs31445079\n",
       "26305 0.000      0.000      0.000      0.000     \n",
       "26306 2.000      2.000      2.000      2.000     \n",
       "26307 2.000      2.000      2.000      2.000     \n",
       "26308 0.968      2.000      0.973      2.000     \n",
       "26309 2.000      2.000      2.000      2.000     \n",
       "26310 2.000      2.000      2.000      2.000     \n",
       "26312 0.000      0.000      0.000      0.000     \n",
       "26313 2.000      2.000      2.000      2.000     \n",
       "26314 2.000      2.000      2.000      2.000     \n",
       "26315 1.000      1.559      1.000      0.561     \n",
       "26316 2.000      2.000      2.000      2.000     \n",
       "26317 2.000      2.000      2.000      2.000     \n",
       "26318 2.000      2.000      2.000      2.000     \n",
       "26319 2.000      2.000      2.000      2.000     \n",
       "26320 2.000      2.000      2.000      2.000     \n",
       "26321 1.031      2.000      1.031      2.000     \n",
       "26322 2.000      2.000      2.000      2.000     \n",
       "26323 2.000      2.000      2.000      2.000     \n",
       "26324 2.000      2.000      2.000      2.000     \n",
       "26325 1.000      1.000      1.000      1.000     \n",
       "26326 2.000      2.000      2.000      2.000     \n",
       "26327 2.000      2.000      2.000      2.000     \n",
       "26328 2.000      2.000      2.000      2.000     \n",
       "26329 0.000      0.000      0.000      0.000     \n",
       "26330 0.000      0.000      0.000      0.000     \n",
       "26331 2.000      2.000      2.000      2.000     \n",
       "26332 2.000      2.000      2.000      2.000     \n",
       "26333 2.000      2.000      2.000      2.000     \n",
       "26334 2.000      2.000      2.000      2.000     \n",
       "26335 2.000      2.000      2.000      2.000     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "X_all[1:30,1:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 10 × 6</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>id</th><th scope=col>chr</th><th scope=col>pos</th><th scope=col>ref</th><th scope=col>alt</th><th scope=col>quality</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>2</th><td>cfw-1-3207478</td><td>1</td><td>3207478</td><td>G</td><td>A</td><td>0.9991</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>cfw-1-4592184</td><td>1</td><td>4592184</td><td>A</td><td>C</td><td>0.9991</td></tr>\n",
       "\t<tr><th scope=row>8</th><td>rs31954814   </td><td>1</td><td>5151352</td><td>T</td><td>G</td><td>0.9931</td></tr>\n",
       "\t<tr><th scope=row>9</th><td>rs31947195   </td><td>1</td><td>5240999</td><td>T</td><td>C</td><td>0.9983</td></tr>\n",
       "\t<tr><th scope=row>10</th><td>rs30660852   </td><td>1</td><td>5241015</td><td>T</td><td>C</td><td>0.9991</td></tr>\n",
       "\t<tr><th scope=row>11</th><td>rs260800880  </td><td>1</td><td>5241052</td><td>C</td><td>G</td><td>0.9974</td></tr>\n",
       "\t<tr><th scope=row>12</th><td>rs49005848   </td><td>1</td><td>5241315</td><td>T</td><td>C</td><td>0.9957</td></tr>\n",
       "\t<tr><th scope=row>13</th><td>rs31616579   </td><td>1</td><td>5259752</td><td>A</td><td>G</td><td>0.9983</td></tr>\n",
       "\t<tr><th scope=row>14</th><td>rs31668389   </td><td>1</td><td>5315021</td><td>T</td><td>C</td><td>0.9974</td></tr>\n",
       "\t<tr><th scope=row>15</th><td>rs31445079   </td><td>1</td><td>5345620</td><td>T</td><td>A</td><td>0.9931</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 10 × 6\n",
       "\\begin{tabular}{r|llllll}\n",
       "  & id & chr & pos & ref & alt & quality\\\\\n",
       "  & <chr> & <fct> & <int> & <fct> & <fct> & <dbl>\\\\\n",
       "\\hline\n",
       "\t2 & cfw-1-3207478 & 1 & 3207478 & G & A & 0.9991\\\\\n",
       "\t6 & cfw-1-4592184 & 1 & 4592184 & A & C & 0.9991\\\\\n",
       "\t8 & rs31954814    & 1 & 5151352 & T & G & 0.9931\\\\\n",
       "\t9 & rs31947195    & 1 & 5240999 & T & C & 0.9983\\\\\n",
       "\t10 & rs30660852    & 1 & 5241015 & T & C & 0.9991\\\\\n",
       "\t11 & rs260800880   & 1 & 5241052 & C & G & 0.9974\\\\\n",
       "\t12 & rs49005848    & 1 & 5241315 & T & C & 0.9957\\\\\n",
       "\t13 & rs31616579    & 1 & 5259752 & A & G & 0.9983\\\\\n",
       "\t14 & rs31668389    & 1 & 5315021 & T & C & 0.9974\\\\\n",
       "\t15 & rs31445079    & 1 & 5345620 & T & A & 0.9931\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 10 × 6\n",
       "\n",
       "| <!--/--> | id &lt;chr&gt; | chr &lt;fct&gt; | pos &lt;int&gt; | ref &lt;fct&gt; | alt &lt;fct&gt; | quality &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|\n",
       "| 2 | cfw-1-3207478 | 1 | 3207478 | G | A | 0.9991 |\n",
       "| 6 | cfw-1-4592184 | 1 | 4592184 | A | C | 0.9991 |\n",
       "| 8 | rs31954814    | 1 | 5151352 | T | G | 0.9931 |\n",
       "| 9 | rs31947195    | 1 | 5240999 | T | C | 0.9983 |\n",
       "| 10 | rs30660852    | 1 | 5241015 | T | C | 0.9991 |\n",
       "| 11 | rs260800880   | 1 | 5241052 | C | G | 0.9974 |\n",
       "| 12 | rs49005848    | 1 | 5241315 | T | C | 0.9957 |\n",
       "| 13 | rs31616579    | 1 | 5259752 | A | G | 0.9983 |\n",
       "| 14 | rs31668389    | 1 | 5315021 | T | C | 0.9974 |\n",
       "| 15 | rs31445079    | 1 | 5345620 | T | A | 0.9931 |\n",
       "\n"
      ],
      "text/plain": [
       "   id            chr pos     ref alt quality\n",
       "2  cfw-1-3207478 1   3207478 G   A   0.9991 \n",
       "6  cfw-1-4592184 1   4592184 A   C   0.9991 \n",
       "8  rs31954814    1   5151352 T   G   0.9931 \n",
       "9  rs31947195    1   5240999 T   C   0.9983 \n",
       "10 rs30660852    1   5241015 T   C   0.9991 \n",
       "11 rs260800880   1   5241052 C   G   0.9974 \n",
       "12 rs49005848    1   5241315 T   C   0.9957 \n",
       "13 rs31616579    1   5259752 A   G   0.9983 \n",
       "14 rs31668389    1   5315021 T   C   0.9974 \n",
       "15 rs31445079    1   5345620 T   A   0.9931 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "map[1:10,]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 10 × 20</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>id</th><th scope=col>round</th><th scope=col>cageid</th><th scope=col>FCbox</th><th scope=col>PPIbox</th><th scope=col>methcage</th><th scope=col>methcycle</th><th scope=col>discard</th><th scope=col>mixup</th><th scope=col>earpunch</th><th scope=col>glucoseage</th><th scope=col>methage</th><th scope=col>FCage</th><th scope=col>PPIage</th><th scope=col>sacage</th><th scope=col>bw0</th><th scope=col>bw1</th><th scope=col>bw2</th><th scope=col>bw3</th><th scope=col>PPIweight</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>4368 </td><td>NA  </td><td>     NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>no</td><td>no</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>  NA</td><td>  NA</td><td>  NA</td><td>  NA</td><td>  NA</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>26305</td><td>SW18</td><td>1330002</td><td>1 </td><td>1 </td><td>1 </td><td>1 </td><td>no</td><td>no</td><td>R </td><td>46</td><td>54</td><td>62</td><td>76</td><td>91</td><td>38.7</td><td>41.0</td><td>41.3</td><td>41.6</td><td>45.7</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>26306</td><td>SW18</td><td>1330002</td><td>2 </td><td>3 </td><td>2 </td><td>1 </td><td>no</td><td>no</td><td>R </td><td>46</td><td>54</td><td>62</td><td>76</td><td>91</td><td>29.1</td><td>29.8</td><td>31.0</td><td>30.6</td><td>35.0</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>26307</td><td>SW18</td><td>1330002</td><td>3 </td><td>4 </td><td>3 </td><td>1 </td><td>no</td><td>no</td><td>L </td><td>46</td><td>54</td><td>62</td><td>76</td><td>91</td><td>28.2</td><td>28.7</td><td>28.4</td><td>29.0</td><td>32.2</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>26308</td><td>SW18</td><td>1330002</td><td>4 </td><td>5 </td><td>4 </td><td>1 </td><td>no</td><td>no</td><td>L </td><td>46</td><td>54</td><td>62</td><td>76</td><td>91</td><td>27.7</td><td>30.6</td><td>31.5</td><td>30.4</td><td>37.5</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>26309</td><td>SW18</td><td>1330003</td><td>1 </td><td>1 </td><td>5 </td><td>1 </td><td>no</td><td>no</td><td>R </td><td>46</td><td>54</td><td>62</td><td>76</td><td>91</td><td>29.1</td><td>31.8</td><td>32.0</td><td>31.9</td><td>37.7</td></tr>\n",
       "\t<tr><th scope=row>7</th><td>26310</td><td>SW18</td><td>1330003</td><td>2 </td><td>3 </td><td>6 </td><td>1 </td><td>no</td><td>no</td><td>R </td><td>46</td><td>54</td><td>62</td><td>76</td><td>91</td><td>30.7</td><td>32.3</td><td>32.2</td><td>32.1</td><td>35.8</td></tr>\n",
       "\t<tr><th scope=row>9</th><td>26312</td><td>SW18</td><td>1330003</td><td>4 </td><td>5 </td><td>8 </td><td>1 </td><td>no</td><td>no</td><td>L </td><td>46</td><td>54</td><td>62</td><td>76</td><td>91</td><td>25.4</td><td>27.8</td><td>27.3</td><td>27.3</td><td>32.1</td></tr>\n",
       "\t<tr><th scope=row>10</th><td>26313</td><td>SW18</td><td>1330004</td><td>1 </td><td>1 </td><td>9 </td><td>1 </td><td>no</td><td>no</td><td>R </td><td>46</td><td>54</td><td>62</td><td>76</td><td>91</td><td>28.1</td><td>30.6</td><td>29.8</td><td>29.8</td><td>33.9</td></tr>\n",
       "\t<tr><th scope=row>12</th><td>26315</td><td>SW18</td><td>1330004</td><td>3 </td><td>4 </td><td>11</td><td>1 </td><td>no</td><td>no</td><td>L </td><td>46</td><td>54</td><td>62</td><td>76</td><td>91</td><td>34.1</td><td>37.3</td><td>37.0</td><td>36.6</td><td>43.7</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 10 × 20\n",
       "\\begin{tabular}{r|llllllllllllllllllll}\n",
       "  & id & round & cageid & FCbox & PPIbox & methcage & methcycle & discard & mixup & earpunch & glucoseage & methage & FCage & PPIage & sacage & bw0 & bw1 & bw2 & bw3 & PPIweight\\\\\n",
       "  & <chr> & <fct> & <int> & <fct> & <fct> & <fct> & <fct> & <fct> & <fct> & <fct> & <int> & <int> & <int> & <int> & <int> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & 4368  & NA   &      NA & NA & NA & NA & NA & no & no & NA & NA & NA & NA & NA & NA &   NA &   NA &   NA &   NA &   NA\\\\\n",
       "\t2 & 26305 & SW18 & 1330002 & 1  & 1  & 1  & 1  & no & no & R  & 46 & 54 & 62 & 76 & 91 & 38.7 & 41.0 & 41.3 & 41.6 & 45.7\\\\\n",
       "\t3 & 26306 & SW18 & 1330002 & 2  & 3  & 2  & 1  & no & no & R  & 46 & 54 & 62 & 76 & 91 & 29.1 & 29.8 & 31.0 & 30.6 & 35.0\\\\\n",
       "\t4 & 26307 & SW18 & 1330002 & 3  & 4  & 3  & 1  & no & no & L  & 46 & 54 & 62 & 76 & 91 & 28.2 & 28.7 & 28.4 & 29.0 & 32.2\\\\\n",
       "\t5 & 26308 & SW18 & 1330002 & 4  & 5  & 4  & 1  & no & no & L  & 46 & 54 & 62 & 76 & 91 & 27.7 & 30.6 & 31.5 & 30.4 & 37.5\\\\\n",
       "\t6 & 26309 & SW18 & 1330003 & 1  & 1  & 5  & 1  & no & no & R  & 46 & 54 & 62 & 76 & 91 & 29.1 & 31.8 & 32.0 & 31.9 & 37.7\\\\\n",
       "\t7 & 26310 & SW18 & 1330003 & 2  & 3  & 6  & 1  & no & no & R  & 46 & 54 & 62 & 76 & 91 & 30.7 & 32.3 & 32.2 & 32.1 & 35.8\\\\\n",
       "\t9 & 26312 & SW18 & 1330003 & 4  & 5  & 8  & 1  & no & no & L  & 46 & 54 & 62 & 76 & 91 & 25.4 & 27.8 & 27.3 & 27.3 & 32.1\\\\\n",
       "\t10 & 26313 & SW18 & 1330004 & 1  & 1  & 9  & 1  & no & no & R  & 46 & 54 & 62 & 76 & 91 & 28.1 & 30.6 & 29.8 & 29.8 & 33.9\\\\\n",
       "\t12 & 26315 & SW18 & 1330004 & 3  & 4  & 11 & 1  & no & no & L  & 46 & 54 & 62 & 76 & 91 & 34.1 & 37.3 & 37.0 & 36.6 & 43.7\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 10 × 20\n",
       "\n",
       "| <!--/--> | id &lt;chr&gt; | round &lt;fct&gt; | cageid &lt;int&gt; | FCbox &lt;fct&gt; | PPIbox &lt;fct&gt; | methcage &lt;fct&gt; | methcycle &lt;fct&gt; | discard &lt;fct&gt; | mixup &lt;fct&gt; | earpunch &lt;fct&gt; | glucoseage &lt;int&gt; | methage &lt;int&gt; | FCage &lt;int&gt; | PPIage &lt;int&gt; | sacage &lt;int&gt; | bw0 &lt;dbl&gt; | bw1 &lt;dbl&gt; | bw2 &lt;dbl&gt; | bw3 &lt;dbl&gt; | PPIweight &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | 4368  | NA   |      NA | NA | NA | NA | NA | no | no | NA | NA | NA | NA | NA | NA |   NA |   NA |   NA |   NA |   NA |\n",
       "| 2 | 26305 | SW18 | 1330002 | 1  | 1  | 1  | 1  | no | no | R  | 46 | 54 | 62 | 76 | 91 | 38.7 | 41.0 | 41.3 | 41.6 | 45.7 |\n",
       "| 3 | 26306 | SW18 | 1330002 | 2  | 3  | 2  | 1  | no | no | R  | 46 | 54 | 62 | 76 | 91 | 29.1 | 29.8 | 31.0 | 30.6 | 35.0 |\n",
       "| 4 | 26307 | SW18 | 1330002 | 3  | 4  | 3  | 1  | no | no | L  | 46 | 54 | 62 | 76 | 91 | 28.2 | 28.7 | 28.4 | 29.0 | 32.2 |\n",
       "| 5 | 26308 | SW18 | 1330002 | 4  | 5  | 4  | 1  | no | no | L  | 46 | 54 | 62 | 76 | 91 | 27.7 | 30.6 | 31.5 | 30.4 | 37.5 |\n",
       "| 6 | 26309 | SW18 | 1330003 | 1  | 1  | 5  | 1  | no | no | R  | 46 | 54 | 62 | 76 | 91 | 29.1 | 31.8 | 32.0 | 31.9 | 37.7 |\n",
       "| 7 | 26310 | SW18 | 1330003 | 2  | 3  | 6  | 1  | no | no | R  | 46 | 54 | 62 | 76 | 91 | 30.7 | 32.3 | 32.2 | 32.1 | 35.8 |\n",
       "| 9 | 26312 | SW18 | 1330003 | 4  | 5  | 8  | 1  | no | no | L  | 46 | 54 | 62 | 76 | 91 | 25.4 | 27.8 | 27.3 | 27.3 | 32.1 |\n",
       "| 10 | 26313 | SW18 | 1330004 | 1  | 1  | 9  | 1  | no | no | R  | 46 | 54 | 62 | 76 | 91 | 28.1 | 30.6 | 29.8 | 29.8 | 33.9 |\n",
       "| 12 | 26315 | SW18 | 1330004 | 3  | 4  | 11 | 1  | no | no | L  | 46 | 54 | 62 | 76 | 91 | 34.1 | 37.3 | 37.0 | 36.6 | 43.7 |\n",
       "\n"
      ],
      "text/plain": [
       "   id    round cageid  FCbox PPIbox methcage methcycle discard mixup earpunch\n",
       "1  4368  NA         NA NA    NA     NA       NA        no      no    NA      \n",
       "2  26305 SW18  1330002 1     1      1        1         no      no    R       \n",
       "3  26306 SW18  1330002 2     3      2        1         no      no    R       \n",
       "4  26307 SW18  1330002 3     4      3        1         no      no    L       \n",
       "5  26308 SW18  1330002 4     5      4        1         no      no    L       \n",
       "6  26309 SW18  1330003 1     1      5        1         no      no    R       \n",
       "7  26310 SW18  1330003 2     3      6        1         no      no    R       \n",
       "9  26312 SW18  1330003 4     5      8        1         no      no    L       \n",
       "10 26313 SW18  1330004 1     1      9        1         no      no    R       \n",
       "12 26315 SW18  1330004 3     4      11       1         no      no    L       \n",
       "   glucoseage methage FCage PPIage sacage bw0  bw1  bw2  bw3  PPIweight\n",
       "1  NA         NA      NA    NA     NA       NA   NA   NA   NA   NA     \n",
       "2  46         54      62    76     91     38.7 41.0 41.3 41.6 45.7     \n",
       "3  46         54      62    76     91     29.1 29.8 31.0 30.6 35.0     \n",
       "4  46         54      62    76     91     28.2 28.7 28.4 29.0 32.2     \n",
       "5  46         54      62    76     91     27.7 30.6 31.5 30.4 37.5     \n",
       "6  46         54      62    76     91     29.1 31.8 32.0 31.9 37.7     \n",
       "7  46         54      62    76     91     30.7 32.3 32.2 32.1 35.8     \n",
       "9  46         54      62    76     91     25.4 27.8 27.3 27.3 32.1     \n",
       "10 46         54      62    76     91     28.1 30.6 29.8 29.8 33.9     \n",
       "12 46         54      62    76     91     34.1 37.3 37.0 36.6 43.7     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pheno_all[1:10,1:20]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Paramaters that can be changed\n",
    "\n",
    "min_var, max_var - which of the columns in genotype data (X_all above) to be considered when using linear models (mostly for speed). Please note that gemma analysis will analyze the whole chromosome\n",
    "\n",
    "analysis_selection - a list of analysis to be processed. For the complete list check src/qtl.analysis.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "analysis_selection <- analyses\n",
    "# min_var = 1\n",
    "# max_var = 100\n",
    "\n",
    "# Example for all variants\n",
    "min_var = 1\n",
    "max_var = dim(X_all)[2]\n",
    "\n",
    "# Example to select only one analysis\n",
    "# analysis_selection <- analyses[\"BMD\"]\n",
    "\n",
    "# Example to select only one chromosome\n",
    "# min_var = min(which(map[\"chr\"]==11))\n",
    "# max_var = max(which(map[\"chr\"]==11))\n",
    "\n",
    "# Examples from sub-graphs (b) and (c) at https://www.nature.com/articles/ng.3609/figures/3 \n",
    "# analysis_selection <- analyses[\"pp12PPIavg\"]\n",
    "# min_var = min(which(map[\"chr\"]==11))\n",
    "# max_var = max(which(map[\"chr\"]==11))\n",
    "\n",
    "# analysis_selection <- analyses[\"pp12PPIavg\"]\n",
    "# min_var = min(which(map[\"chr\"]==7))\n",
    "# max_var = max(which(map[\"chr\"]==7))\n",
    "\n",
    "# analysis_selection <- analyses[\"testis\"]\n",
    "# min_var = min(which(map[\"chr\"]==13))\n",
    "# max_var = max(which(map[\"chr\"]==13))\n",
    "\n",
    "# analysis_selection <- analyses[\"testis\"]\n",
    "# min_var = min(which(map[\"chr\"]==5))\n",
    "# max_var = max(which(map[\"chr\"]==5))\n",
    "\n",
    "chromosomes <- unique(map[min_var:max_var,\"chr\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "For example usage please run: vignette('qqman')\n",
      "\n",
      "\n",
      "\n",
      "Citation appreciated but not required:\n",
      "\n",
      "Turner, (2018). qqman: an R package for visualizing GWAS results using Q-Q and manhattan plots. Journal of Open Source Software, 3(25), 731, https://doi.org/10.21105/joss.00731.\n",
      "\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "68 outliers for BMD conditioned on SW16 are removed.\n",
      "Mapping QTLs for BMD in 1038 mice, controlling for SW16.\n",
      "Writing phenotype and covariate data to file.\n",
      "Mapping QTLs on chromosome 1.\n",
      " * Computing kinship matrix.\n",
      " * Writing kinship matrix to file.\n",
      " * Writing to file genotypes for 6372 markers on chromosome 1.\n",
      " * Writing genetic map for 6372 markers on chromosome 1 to file.\n",
      " * Computing p-values for 6372 markers on chromosome 1.\n",
      "Mapping QTLs on chromosome 2.\n",
      " * Computing kinship matrix.\n",
      " * Writing kinship matrix to file.\n",
      " * Writing to file genotypes for 7896 markers on chromosome 2.\n",
      " * Writing genetic map for 7896 markers on chromosome 2 to file.\n",
      " * Computing p-values for 7896 markers on chromosome 2.\n",
      "Mapping QTLs on chromosome 3.\n",
      " * Computing kinship matrix.\n",
      " * Writing kinship matrix to file.\n",
      " * Writing to file genotypes for 2848 markers on chromosome 3.\n",
      " * Writing genetic map for 2848 markers on chromosome 3 to file.\n",
      " * Computing p-values for 2848 markers on chromosome 3.\n",
      "Mapping QTLs on chromosome 4.\n",
      " * Computing kinship matrix.\n",
      " * Writing kinship matrix to file.\n",
      " * Writing to file genotypes for 5201 markers on chromosome 4.\n",
      " * Writing genetic map for 5201 markers on chromosome 4 to file.\n",
      " * Computing p-values for 5201 markers on chromosome 4.\n",
      "Mapping QTLs on chromosome 5.\n",
      " * Computing kinship matrix.\n",
      " * Writing kinship matrix to file.\n",
      " * Writing to file genotypes for 4707 markers on chromosome 5.\n",
      " * Writing genetic map for 4707 markers on chromosome 5 to file.\n",
      " * Computing p-values for 4707 markers on chromosome 5.\n",
      "Mapping QTLs on chromosome 6.\n",
      " * Computing kinship matrix.\n",
      " * Writing kinship matrix to file.\n",
      " * Writing to file genotypes for 5477 markers on chromosome 6.\n",
      " * Writing genetic map for 5477 markers on chromosome 6 to file.\n",
      " * Computing p-values for 5477 markers on chromosome 6.\n",
      "Mapping QTLs on chromosome 7.\n",
      " * Computing kinship matrix.\n",
      " * Writing kinship matrix to file.\n",
      " * Writing to file genotypes for 3625 markers on chromosome 7.\n",
      " * Writing genetic map for 3625 markers on chromosome 7 to file.\n",
      " * Computing p-values for 3625 markers on chromosome 7.\n",
      "Mapping QTLs on chromosome 8.\n",
      " * Computing kinship matrix.\n",
      " * Writing kinship matrix to file.\n",
      " * Writing to file genotypes for 3574 markers on chromosome 8.\n",
      " * Writing genetic map for 3574 markers on chromosome 8 to file.\n",
      " * Computing p-values for 3574 markers on chromosome 8.\n",
      "Mapping QTLs on chromosome 9.\n",
      " * Computing kinship matrix.\n",
      " * Writing kinship matrix to file.\n",
      " * Writing to file genotypes for 3625 markers on chromosome 9.\n",
      " * Writing genetic map for 3625 markers on chromosome 9 to file.\n",
      " * Computing p-values for 3625 markers on chromosome 9.\n",
      "Mapping QTLs on chromosome 10.\n",
      " * Computing kinship matrix.\n",
      " * Writing kinship matrix to file.\n",
      " * Writing to file genotypes for 2700 markers on chromosome 10.\n",
      " * Writing genetic map for 2700 markers on chromosome 10 to file.\n",
      " * Computing p-values for 2700 markers on chromosome 10.\n",
      "Mapping QTLs on chromosome 11.\n",
      " * Computing kinship matrix.\n",
      " * Writing kinship matrix to file.\n",
      " * Writing to file genotypes for 6624 markers on chromosome 11.\n",
      " * Writing genetic map for 6624 markers on chromosome 11 to file.\n",
      " * Computing p-values for 6624 markers on chromosome 11.\n",
      "Mapping QTLs on chromosome 12.\n",
      " * Computing kinship matrix.\n",
      " * Writing kinship matrix to file.\n",
      " * Writing to file genotypes for 5798 markers on chromosome 12.\n",
      " * Writing genetic map for 5798 markers on chromosome 12 to file.\n",
      " * Computing p-values for 5798 markers on chromosome 12.\n",
      "Mapping QTLs on chromosome 13.\n",
      " * Computing kinship matrix.\n",
      " * Writing kinship matrix to file.\n",
      " * Writing to file genotypes for 4004 markers on chromosome 13.\n",
      " * Writing genetic map for 4004 markers on chromosome 13 to file.\n",
      " * Computing p-values for 4004 markers on chromosome 13.\n",
      "Mapping QTLs on chromosome 14.\n",
      " * Computing kinship matrix.\n",
      " * Writing kinship matrix to file.\n",
      " * Writing to file genotypes for 4146 markers on chromosome 14.\n",
      " * Writing genetic map for 4146 markers on chromosome 14 to file.\n",
      " * Computing p-values for 4146 markers on chromosome 14.\n",
      "Mapping QTLs on chromosome 15.\n",
      " * Computing kinship matrix.\n",
      " * Writing kinship matrix to file.\n",
      " * Writing to file genotypes for 3596 markers on chromosome 15.\n",
      " * Writing genetic map for 3596 markers on chromosome 15 to file.\n",
      " * Computing p-values for 3596 markers on chromosome 15.\n",
      "Mapping QTLs on chromosome 16.\n",
      " * Computing kinship matrix.\n",
      " * Writing kinship matrix to file.\n",
      " * Writing to file genotypes for 741 markers on chromosome 16.\n",
      " * Writing genetic map for 741 markers on chromosome 16 to file.\n",
      " * Computing p-values for 741 markers on chromosome 16.\n",
      "Mapping QTLs on chromosome 17.\n",
      " * Computing kinship matrix.\n",
      " * Writing kinship matrix to file.\n",
      " * Writing to file genotypes for 4671 markers on chromosome 17.\n",
      " * Writing genetic map for 4671 markers on chromosome 17 to file.\n",
      " * Computing p-values for 4671 markers on chromosome 17.\n",
      "Mapping QTLs on chromosome 18.\n",
      " * Computing kinship matrix.\n",
      " * Writing kinship matrix to file.\n",
      " * Writing to file genotypes for 2213 markers on chromosome 18.\n",
      " * Writing genetic map for 2213 markers on chromosome 18 to file.\n",
      " * Computing p-values for 2213 markers on chromosome 18.\n",
      "Mapping QTLs on chromosome 19.\n",
      " * Computing kinship matrix.\n",
      " * Writing kinship matrix to file.\n",
      " * Writing to file genotypes for 1828 markers on chromosome 19.\n",
      " * Writing genetic map for 1828 markers on chromosome 19 to file.\n",
      " * Computing p-values for 1828 markers on chromosome 19.\n",
      "Saving results to file.\n",
      "[1] 79646\n",
      "[1] \"Manhattan plot output to: out/lm_BMD_1_79646.png, (sorted) values saved in: out/lm_BMD_1_79646.csv\"\n"
     ]
    }
   ],
   "source": [
    "library(qqman)\n",
    "library(data.table)\n",
    "\n",
    "for(analysis in analysis_selection) { \n",
    "\n",
    "    ##################################\n",
    "    # Cleanup data\n",
    "    phenotype  <- analysis$pheno\n",
    "    covariates <- analysis$cov\n",
    "    outliers   <- analysis$outliers\n",
    "    \n",
    "    pheno <- copy(pheno_all)\n",
    "    if (!is.null(outliers))\n",
    "      pheno_all <- remove.outliers(pheno,phenotype,covariates,outliers)\n",
    "\n",
    "    \n",
    "    # Only analyze samples (i.e. rows of the genotype and phenotype\n",
    "    # matrices) for which the phenotype and all the covariates are\n",
    "    # observed.\n",
    "    pheno <- pheno[which(none.missing.row(pheno[c(phenotype,covariates)])),]\n",
    "\n",
    "    # Align the phenotypes and genotypes\n",
    "    ids   <- intersect(pheno_all$id,rownames(X_all))\n",
    "    pheno <- pheno_all[match(ids,pheno_all$id),]\n",
    "    X     <- X_all[match(ids,rownames(X_all)),]\n",
    "\n",
    "    ###################################\n",
    "    # Compute using gemma\n",
    "    # MAP QTLs \n",
    "    ge_out_dat <- paste(\"out/ge_\", analysis$pheno, \"_\", min_var, \"_\", max_var, \".dat\",sep=\"\")\n",
    "    ge_out_csv <- paste(\"out/ge_\", analysis$pheno, \"_\", min_var, \"_\", max_var, \".csv\",sep=\"\")\n",
    "    ge_out_png <- paste(\"out/ge_\", analysis$pheno, \"_\", min_var, \"_\", max_var, \".png\",sep=\"\")\n",
    "    \n",
    "    if (!file.exists(ge_out_csv) | !file.exists(ge_out_png)) {\n",
    "        # Calculate p-values using GEMMA.\n",
    "        gwscan.gemma <- run.gemma(phenotype,covariates,pheno,X,map,\n",
    "                                  gemmadir,gemma.exe,chromosomes)\n",
    "\n",
    "        # Save results to file.\n",
    "        cat(\"Saving results to file.\\n\")\n",
    "        save(list = c(\"analysis\",\"gwscan.gemma\"),file = ge_out_dat)\n",
    "        \n",
    "        named_gws <- gwscan.gemma\n",
    "        named_gws$snp = rownames(named_gws)\n",
    "        named_gws$p = 10 ^ (-named_gws$log10p)\n",
    "        \n",
    "        png(filename = ge_out_png,\n",
    "            width = 600, height = 600, units = \"px\", pointsize = 12,\n",
    "             bg = \"white\",  res = NA,\n",
    "             type = c(\"cairo\", \"cairo-png\", \"Xlib\", \"quartz\"))\n",
    "        manhattan(named_gws, chr=\"chr\", bp=\"pos\", snp='snp', p=\"p\", annotatePval = 0.1, annotateTop = TRUE )\n",
    "        dev.off()\n",
    "        \n",
    "        write.csv(data.table(named_gws)[order(rank(p)),], ge_out_csv)\n",
    "        \n",
    "    }\n",
    "    \n",
    "    ###################################\n",
    "    # Compute using linear model\n",
    "    lm_out_csv <- paste(\"out/lm_\", analysis$pheno, \"_\", min_var, \"_\", max_var, \".csv\",sep=\"\")\n",
    "    lm_out_png <- paste(\"out/lm_\", analysis$pheno, \"_\", min_var, \"_\", max_var, \".png\",sep=\"\")\n",
    "    \n",
    "    if(!file.exists(lm_out_csv) | ! file.exists(lm_out_png)) {\n",
    "        print(dim(X)[2])\n",
    "        dt <- data.table(snp=rep(\"\",dim(X)[2]), chr=rep(0,dim(X)[2]), pos=rep(0,dim(X)[2]), p=rep(1,dim(X)[2]))\n",
    "        for (i in min_var:max_var) {\n",
    "            X_variant <- cbind(X[,i], pheno_column=pheno[,analysis$pheno])\n",
    "            colnames(X_variant)[1]<-colnames(X)[i]\n",
    "            f <- paste(\"pheno_column ~ \",colnames(X)[i])\n",
    "            # Add any covariates\n",
    "            for(cov in analysis$cov) {\n",
    "                X_variant <- cbind(X_variant, pheno_column=pheno[,cov])\n",
    "                f <- paste(f,\"+\",cov)\n",
    "            }\n",
    "            res_variant <- lm(pheno_column~., data = data.frame(X_variant))\n",
    "            dt[i,1] = colnames(X)[i]\n",
    "            dt[i,2] = as.numeric(map[map[\"id\"]==colnames(X)[i],\"chr\"])\n",
    "            dt[i,3] = as.numeric(map[map[\"id\"]==colnames(X)[i],\"pos\"])\n",
    "            dt[i,4] = as.numeric(summary(res_variant)$coefficients[2,4])\n",
    "        }\n",
    "\n",
    "        # Print to file\n",
    "        png(filename = lm_out_png,\n",
    "            width = 600, height = 600, units = \"px\", pointsize = 12,\n",
    "             bg = \"white\",  res = NA,\n",
    "             type = c(\"cairo\", \"cairo-png\", \"Xlib\", \"quartz\"))\n",
    "        manhattan(dt[min_var:max_var], chr=\"chr\", bp=\"pos\", snp='snp', p=\"p\", annotatePval = 0.1, annotateTop = TRUE)\n",
    "        write.csv(dt[order(rank(p)),][1:(max_var-min_var)], lm_out_csv)\n",
    "        dev.off()\n",
    "        print(paste(\"Manhattan plot output to: \", lm_out_png, \", (sorted) values saved in: \", lm_out_csv, sep=\"\"))\n",
    "        ## Print also on screen\n",
    "        ## manhattan(dt[min_var:max_var], chr=\"chr\", bp=\"pos\", snp='snp', p=\"p\", annotatePval = 0.1, annotateTop = TRUE)\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.2.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
