{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Build the mouse genome training dataset\n",
    "This notebook creates a mouse genome training datasets by batching the genome data by position on the chromosome. You can pick the quantity and characteristics of the batches you'd like to create training \n",
    "datasets for. Your choices are written to a file that the next notebook, 04_create_synthetic_mouse_genomes, reads from."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pathlib\n",
    "import pandas as pd\n",
    "\n",
    "base_path = pathlib.Path(os.getcwd().replace(\"/synthetics\", \"\"))\n",
    "data_path = base_path / 'mice_data_set' / 'data' \n",
    "experiment_path = base_path / 'mice_data_set' / 'out' "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/amy/.pyenv/versions/3.7.2/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3170: DtypeWarning: Columns (0) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>discard</th>\n",
       "      <th>cfw-1-3082859</th>\n",
       "      <th>cfw-1-3207478</th>\n",
       "      <th>cfw-1-3284999</th>\n",
       "      <th>cfw-1-4056451</th>\n",
       "      <th>rs241840178</th>\n",
       "      <th>cfw-1-4592184</th>\n",
       "      <th>rs214108183</th>\n",
       "      <th>rs31954814</th>\n",
       "      <th>...</th>\n",
       "      <th>rs239202862</th>\n",
       "      <th>cfw-19-60773695</th>\n",
       "      <th>rs212272420</th>\n",
       "      <th>rs51223003</th>\n",
       "      <th>rs30736750</th>\n",
       "      <th>rs30654044</th>\n",
       "      <th>rs30990073</th>\n",
       "      <th>rs50978457</th>\n",
       "      <th>rs51755773</th>\n",
       "      <th>cfw-19-61107432</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>26305</td>\n",
       "      <td>no</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.000</td>\n",
       "      <td>...</td>\n",
       "      <td>2.000</td>\n",
       "      <td>2.000</td>\n",
       "      <td>2.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.000</td>\n",
       "      <td>2.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>26306</td>\n",
       "      <td>no</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>...</td>\n",
       "      <td>1.929</td>\n",
       "      <td>1.980</td>\n",
       "      <td>1.999</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.964</td>\n",
       "      <td>1.689</td>\n",
       "      <td>1.781</td>\n",
       "      <td>0.254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>26307</td>\n",
       "      <td>no</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>...</td>\n",
       "      <td>1.933</td>\n",
       "      <td>1.603</td>\n",
       "      <td>1.993</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.492</td>\n",
       "      <td>1.793</td>\n",
       "      <td>1.538</td>\n",
       "      <td>0.526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>26308</td>\n",
       "      <td>no</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.776</td>\n",
       "      <td>...</td>\n",
       "      <td>2.000</td>\n",
       "      <td>2.000</td>\n",
       "      <td>2.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.000</td>\n",
       "      <td>2.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>26309</td>\n",
       "      <td>no</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>...</td>\n",
       "      <td>2.000</td>\n",
       "      <td>2.000</td>\n",
       "      <td>2.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.000</td>\n",
       "      <td>2.000</td>\n",
       "      <td>2.000</td>\n",
       "      <td>0.361</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 92736 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      id discard  cfw-1-3082859  cfw-1-3207478  cfw-1-3284999  cfw-1-4056451  \\\n",
       "0  26305      no            0.0            0.0            0.0            0.0   \n",
       "1  26306      no            0.0            0.0            0.0            0.0   \n",
       "2  26307      no            0.0            0.0            0.0            0.0   \n",
       "3  26308      no            0.0            0.0            0.0            0.0   \n",
       "4  26309      no            0.0            0.0            0.0            0.0   \n",
       "\n",
       "   rs241840178  cfw-1-4592184  rs214108183  rs31954814  ...  rs239202862  \\\n",
       "0          0.0            0.0          0.0       2.000  ...        2.000   \n",
       "1          0.0            0.0          0.0       0.000  ...        1.929   \n",
       "2          0.0            2.0          0.0       0.000  ...        1.933   \n",
       "3          0.0            0.0          0.0       0.776  ...        2.000   \n",
       "4          0.0            0.0          0.0       0.000  ...        2.000   \n",
       "\n",
       "   cfw-19-60773695  rs212272420  rs51223003  rs30736750  rs30654044  \\\n",
       "0            2.000        2.000         0.0         0.0         2.0   \n",
       "1            1.980        1.999         0.0         0.0         2.0   \n",
       "2            1.603        1.993         0.0         0.0         2.0   \n",
       "3            2.000        2.000         0.0         0.0         2.0   \n",
       "4            2.000        2.000         0.0         0.0         2.0   \n",
       "\n",
       "   rs30990073  rs50978457  rs51755773  cfw-19-61107432  \n",
       "0       2.000       2.000       0.000            0.000  \n",
       "1       1.964       1.689       1.781            0.254  \n",
       "2       1.492       1.793       1.538            0.526  \n",
       "3       2.000       2.000       1.000            0.000  \n",
       "4       2.000       2.000       2.000            0.361  \n",
       "\n",
       "[5 rows x 92736 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read in the geno data and remove the discards (slow)\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "genofile = data_path / \"geno.txt\"\n",
    "geno = pd.read_csv(genofile, sep=' ')\n",
    "geno = geno[geno[\"discard\"] == \"no\"]\n",
    "geno.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TA</th>\n",
       "      <th>SW16</th>\n",
       "      <th>tibia</th>\n",
       "      <th>EDL</th>\n",
       "      <th>plantaris</th>\n",
       "      <th>gastroc</th>\n",
       "      <th>SW6</th>\n",
       "      <th>sacweight</th>\n",
       "      <th>BMD</th>\n",
       "      <th>abBMD</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>74.6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>19.05</td>\n",
       "      <td>16.9</td>\n",
       "      <td>23.2</td>\n",
       "      <td>194.8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>46.6</td>\n",
       "      <td>1.92</td>\n",
       "      <td>0.0</td>\n",
       "      <td>26305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>62.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>18.06</td>\n",
       "      <td>13.2</td>\n",
       "      <td>18.8</td>\n",
       "      <td>154.6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>35.7</td>\n",
       "      <td>1.88</td>\n",
       "      <td>0.0</td>\n",
       "      <td>26306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>54.1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>18.14</td>\n",
       "      <td>11.2</td>\n",
       "      <td>17.6</td>\n",
       "      <td>143.6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>34.1</td>\n",
       "      <td>1.89</td>\n",
       "      <td>0.0</td>\n",
       "      <td>26307</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>56.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>18.18</td>\n",
       "      <td>12.9</td>\n",
       "      <td>17.7</td>\n",
       "      <td>148.6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>41.8</td>\n",
       "      <td>1.95</td>\n",
       "      <td>0.0</td>\n",
       "      <td>26308</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>64.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>18.35</td>\n",
       "      <td>14.1</td>\n",
       "      <td>20.6</td>\n",
       "      <td>157.4</td>\n",
       "      <td>0.0</td>\n",
       "      <td>39.5</td>\n",
       "      <td>1.92</td>\n",
       "      <td>0.0</td>\n",
       "      <td>26309</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     TA  SW16  tibia   EDL  plantaris  gastroc  SW6  sacweight   BMD  abBMD  \\\n",
       "0  74.6   0.0  19.05  16.9       23.2    194.8  0.0       46.6  1.92    0.0   \n",
       "1  62.3   0.0  18.06  13.2       18.8    154.6  0.0       35.7  1.88    0.0   \n",
       "2  54.1   0.0  18.14  11.2       17.6    143.6  0.0       34.1  1.89    0.0   \n",
       "3  56.5   0.0  18.18  12.9       17.7    148.6  0.0       41.8  1.95    0.0   \n",
       "4  64.0   0.0  18.35  14.1       20.6    157.4  0.0       39.5  1.92    0.0   \n",
       "\n",
       "      id  \n",
       "0  26305  \n",
       "1  26306  \n",
       "2  26307  \n",
       "3  26308  \n",
       "4  26309  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read in the muscle and bone phenome data, which contains the abBMD phenome and its covariate\n",
    "\n",
    "phenofile = data_path / \"pheno_batch0_withID.csv\"\n",
    "pheno = pd.read_csv(phenofile)\n",
    "pheno.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>snp</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>p</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>rs29477109</td>\n",
       "      <td>11</td>\n",
       "      <td>95292217</td>\n",
       "      <td>5.052317e-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>rs27071351</td>\n",
       "      <td>11</td>\n",
       "      <td>96114911</td>\n",
       "      <td>7.074181e-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>rs27024162</td>\n",
       "      <td>11</td>\n",
       "      <td>96918116</td>\n",
       "      <td>7.170582e-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>rs49423067</td>\n",
       "      <td>11</td>\n",
       "      <td>96918212</td>\n",
       "      <td>7.198661e-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>rs29470802</td>\n",
       "      <td>11</td>\n",
       "      <td>95263588</td>\n",
       "      <td>8.049849e-14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0         snp  chr       pos             p\n",
       "0           1  rs29477109   11  95292217  5.052317e-14\n",
       "1           2  rs27071351   11  96114911  7.074181e-14\n",
       "2           3  rs27024162   11  96918116  7.170582e-14\n",
       "3           4  rs49423067   11  96918212  7.198661e-14\n",
       "4           5  rs29470802   11  95263588  8.049849e-14"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Grab the original GWAS linear model results for abBMD and the chromosome position information\n",
    "\n",
    "gwasfile = experiment_path / \"lm_abBMD_1_79646.csv\"\n",
    "gwas_scores = pd.read_csv(gwasfile)\n",
    "gwas_scores.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>snp</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>p</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>73583</td>\n",
       "      <td>73584</td>\n",
       "      <td>cfw-1-3207478</td>\n",
       "      <td>1</td>\n",
       "      <td>3207478</td>\n",
       "      <td>0.921913</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>40919</td>\n",
       "      <td>40920</td>\n",
       "      <td>cfw-1-4592184</td>\n",
       "      <td>1</td>\n",
       "      <td>4592184</td>\n",
       "      <td>0.496735</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>29303</td>\n",
       "      <td>29304</td>\n",
       "      <td>rs31954814</td>\n",
       "      <td>1</td>\n",
       "      <td>5151352</td>\n",
       "      <td>0.353184</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>44285</td>\n",
       "      <td>44286</td>\n",
       "      <td>rs31947195</td>\n",
       "      <td>1</td>\n",
       "      <td>5240999</td>\n",
       "      <td>0.540256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>40335</td>\n",
       "      <td>40336</td>\n",
       "      <td>rs30660852</td>\n",
       "      <td>1</td>\n",
       "      <td>5241015</td>\n",
       "      <td>0.489395</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   index  Unnamed: 0            snp  chr      pos         p\n",
       "0  73583       73584  cfw-1-3207478    1  3207478  0.921913\n",
       "1  40919       40920  cfw-1-4592184    1  4592184  0.496735\n",
       "2  29303       29304     rs31954814    1  5151352  0.353184\n",
       "3  44285       44286     rs31947195    1  5240999  0.540256\n",
       "4  40335       40336     rs30660852    1  5241015  0.489395"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Sort by position on the chromosome\n",
    "\n",
    "gwas_scores_sorted = gwas_scores.sort_values(by=['chr','pos']).reset_index()\n",
    "gwas_scores_sorted.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Group the SNPs into batches of 17 (saving room to use abBMD and SW16 as seeds)\n",
    "\n",
    "interesting_threshold = 5e-8\n",
    "\n",
    "batches = {}\n",
    "batch_num = -1\n",
    "max_snps_per_batch = 17\n",
    "snp_cnt = 0\n",
    "last_chromo = -1\n",
    "batch_psum = {}\n",
    "batch_avg_pvalue = {}\n",
    "all_avg_pvalues = []\n",
    "all_int_cnt = []\n",
    "all_not_int_cnt = []\n",
    "all_batch_nums = []\n",
    "\n",
    "for i in range(len(gwas_scores_sorted)):\n",
    "    pscore = gwas_scores_sorted.loc[i]['p']\n",
    "    chromo = gwas_scores_sorted.loc[i]['chr']\n",
    "    pos = gwas_scores_sorted.loc[i]['pos'] \n",
    "    snp = gwas_scores_sorted.loc[i]['snp'] \n",
    "    if ((snp_cnt == max_snps_per_batch) or (chromo != last_chromo)):\n",
    "        batch_num += 1\n",
    "        batches[batch_num] = {}\n",
    "        batches[batch_num]['chr_pos'] = []\n",
    "        batches[batch_num]['pvalues'] = []\n",
    "        batches[batch_num]['snps'] = []\n",
    "        batches[batch_num]['snp_cnt'] = 0\n",
    "        batches[batch_num]['interesting_cnt'] = 0\n",
    "        batches[batch_num]['not_interesting_cnt'] = 0       \n",
    "        batch_psum[batch_num] = 0\n",
    "        snp_cnt = 0\n",
    "    batches[batch_num]['chr_pos'].append(str(chromo) + \"_\" + str(pos))\n",
    "    batches[batch_num]['pvalues'].append(pscore)\n",
    "    batches[batch_num]['snps'].append(snp)\n",
    "    batches[batch_num]['snp_cnt'] += 1\n",
    "    if pscore <= interesting_threshold:\n",
    "        batches[batch_num]['interesting_cnt'] += 1\n",
    "    else:\n",
    "        batches[batch_num]['not_interesting_cnt'] += 1\n",
    "    batch_psum[batch_num] += pscore\n",
    "    last_chromo = chromo\n",
    "    snp_cnt += 1\n",
    "    \n",
    "# Compute the average pvalue per batch\n",
    "for i, next_batch in enumerate(batches):\n",
    "    batch_avg_pvalue[next_batch] = batch_psum[next_batch] / batches[next_batch]['snp_cnt']\n",
    "    all_avg_pvalues.append(batch_avg_pvalue[next_batch])\n",
    "    all_int_cnt.append(batches[next_batch]['interesting_cnt'])\n",
    "    all_not_int_cnt.append(batches[next_batch]['not_interesting_cnt'])\n",
    "    all_batch_nums.append(i)\n",
    "    \n",
    "pvalue_df = pd.DataFrame({\"batch\": all_batch_nums, \"avg_pvalue\": all_avg_pvalues, \"interesting_cnt\": all_int_cnt, \"not_interesting_cnt\": all_not_int_cnt})\n",
    "     "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create function to build a training set from a batch\n",
    "\n",
    "def build_training(batch):\n",
    "    \n",
    "    training_min_rows = 25000\n",
    "    \n",
    "    # Gather the SNPs\n",
    "    grp_columns = list(batches[batch][\"snps\"])\n",
    "    grp_columns.append(\"id\")\n",
    "    geno_grp = geno.filter(grp_columns)\n",
    "\n",
    "    # Cast float values to integers\n",
    "    floats = geno_grp.select_dtypes(include=['float64']) \n",
    "    for col in floats.columns.values:\n",
    "        geno_grp[col] = geno_grp[col].astype('int64')\n",
    "\n",
    "    # Add in the phenome information to the genome training set\n",
    "    genome_phenome = geno_grp.join(pheno.set_index('id'), on = \"id\", how = \"inner\")\n",
    "    columns_use = list(batches[batch][\"snps\"])\n",
    "    columns_use.append(\"abBMD\")\n",
    "    columns_use.append(\"SW16\")\n",
    "    genome_train = genome_phenome.filter(columns_use)\n",
    "    \n",
    "    # Replicate training set to have a minimum of 25000 examples\n",
    "    dataset_rows = len(genome_train)\n",
    "    genome_train = pd.concat([genome_train] * (training_min_rows // dataset_rows + 1))\n",
    "    \n",
    "    # Save the training file\n",
    "    filename = \"geno_abBMD_batch\" + str(batch) + \"_train.csv\"\n",
    "    genofile = data_path / \"genome_training_data\" / filename\n",
    "    genome_train.to_csv(genofile, index=False, header=True)\n",
    "    \n",
    "    # Now create of version of map.txt with just the SNPs in the training set of this first batch\n",
    "    mapfile = data_path / \"map.txt\"\n",
    "    mapdata = pd.read_csv(mapfile, sep=' ')\n",
    "    mapdata_use = mapdata[mapdata[\"id\"].isin(batches[batch][\"snps\"])]\n",
    "    filename = \"map_abBMD_batch\" + str(batch) + \".txt\"\n",
    "    mapfile_new = data_path / \"genome_map_data\" / filename\n",
    "    mapdata_use.to_csv(mapfile_new, sep=' ', header=True, index=False)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "40"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# How many batches have interesting pvalues\n",
    "len(pvalue_df[pvalue_df[\"interesting_cnt\"] > 0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create training sets for some or all of the batches with interesting pvalues\n",
    "\n",
    "interesting_batches_to_use = 40\n",
    "all_batches_used = []\n",
    "batches_to_use = list(pvalue_df[pvalue_df[\"interesting_cnt\"] > 0][\"batch\"].sample(n=interesting_batches_to_use))\n",
    "for batch in batches_to_use:\n",
    "    #build_training(batch)\n",
    "    all_batches_used.append(batch)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4654"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# How many batches have nothing interesting\n",
    "len(pvalue_df[pvalue_df[\"interesting_cnt\"] == 0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create training sets for some or all of the batches with no interesting pvalues\n",
    "\n",
    "non_interesting_batches_to_use = 500\n",
    "batches_to_use = list(pvalue_df[pvalue_df[\"interesting_cnt\"] == 0][\"batch\"].sample(n=non_interesting_batches_to_use))\n",
    "for batch in batches_to_use:\n",
    "    #build_training(batch)\n",
    "    all_batches_used.append(batch)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save to a file the list of batch numbers we created training sets for\n",
    "\n",
    "file_df = pd.DataFrame({\"batch\": all_batches_used})\n",
    "filename = data_path / \"batch_training_list.csv\"\n",
    "file_df.to_csv(filename, index=False, header=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Additionally, create a batch file list containing all batches\n",
    "\n",
    "all_batches = list(pvalue_df[\"batch\"])\n",
    "for batch in all_batches:\n",
    "    if batch not in all_batches_used:\n",
    "        build_training(batch)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4694"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(all_batches)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save to a file this list of all batches\n",
    "\n",
    "file_df = pd.DataFrame({\"batch\": all_batches})\n",
    "filename = data_path / \"batch_training_list_all.csv\"\n",
    "file_df.to_csv(filename, index=False, header=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
